python/morpheus/morpheus/utils/shared_process_pool.py

```diff
@@ -0,0 +1,409 @@
+        task_queue_count = len(self._task_queues)
+        empty_task_queues = set()
+        while len(empty_task_queues) < task_queue_count:
+            for stage_name, task_queue in self._task_queues.items():
+                if task_queue.empty():
+                    empty_task_queues.add(stage_name)
+
+        self._cancellation_token.cancel()
+
+        for i, p in enumerate(self._processes):
+            p.join()
+            logger.debug("Process %s/%s has been joined.", i + 1, self._total_max_workers)
+
+        logger.debug("SharedProcessPool.stop(): All tasks have been completed. SharedProcessPool has been stopped.")
+        self._status = PoolStatus.SHUTDOWN
+
+    def terminate(self):
+        """
+        Terminate all processes and shutdown the SharedProcessPool immediately.
+        """
+        for i, p in enumerate(self._processes):
+            p.terminate()
+            logger.debug("Process %s/%s has been terminated.", i + 1, self._total_max_workers)
+
+        logger.debug("SharedProcessPool.terminate(): SharedProcessPool has been terminated.")
+        self._status = PoolStatus.SHUTDOWN
+
+    def wait_until_ready(self, timeout=None):
+        """
+        Wait until the SharedProcessPool is running and ready to accept tasks.
+
+        Parameters
+        ----------
+        timeout : _type_, optional
+            timeout in seconds to wait for the SharedProcessPool to be ready, by default None.
+            If None, it will wait indefinitely.
+
+        Raises
+        ------
+        RuntimeError
+            If the SharedProcessPool is not initializing or running.
+        TimeoutError
+            If has been waiting more than the timeout.
+        """
+        if self.status not in (PoolStatus.INITIALIZING, PoolStatus.RUNNING):
+            raise RuntimeError("Cannot wait for a SharedProcessPool that is not initializing.")
+
+        start_time = time.time()
+        while self.status != PoolStatus.RUNNING:
+            if timeout is not None and time.time() - start_time > timeout:
+                raise TimeoutError("SharedProcessPool wait_until_ready has timed out.")
+            time.sleep(0.1)
+
+        logger.debug("SharedProcessPool.wait_until_ready(): SharedProcessPool is ready.")
+
+    def join(self, timeout=None):
+        """
+        Wait until the SharedProcessPool is terminated.
+
+        Parameters
+        ----------
+        timeout : _type_, optional
+            timeout in seconds to wait for the SharedProcessPool to be terminated, by default None.
+            If None, it will wait indefinitely.
+
+        Raises
+        ------
+        TimeoutError
+            If has been waiting more than the timeout.
+        """
+        start_time = time.time()
+
+        while self._status != PoolStatus.SHUTDOWN:
+            if timeout is not None and time.time() - start_time > timeout:
+                raise TimeoutError("SharedProcessPool join has timed out.")
+            time.sleep(0.1)
```

Comment:
We need to improve this join method. A better way would be to use some other synchronization primitive. Usually, a condition variable with a predicate works well here. Let me know if you need examples. When using a condition variable, you will need to spawn a new thread which does something like:
```python
for p in self._processes:
   p.join()

self._status = PoolStatus.SHUTDOWN

self._condition_variable.notify_all()
```